torchio/transforms/augmentation/intensity/random_labels_to_image.py

```diff
@@ -0,0 +1,256 @@
+
+from typing import Union, Tuple, Optional, Dict, Sequence
+import torch
+import numpy as np
+from ....torchio import DATA, TypeData, TypeRangeFloat, TypeNumber, AFFINE, INTENSITY
+from ....data.subject import Subject
+from ....data.image import Image
+from .. import RandomTransform
+
+MEAN_RANGE = (0.1, 0.9)
+STD_RANGE = (0.01, 0.1)
+
+
+class RandomLabelsToImage(RandomTransform):
+    r"""Generate a random image from a binary label map or a set of
+     partial volume (PV) label maps.
+
+    Args:
+        label_key: String designating the label map in the sample
+            that will be used to generate the new image.
+            Cannot be set at the same time as pv_label_keys.
+        pv_label_keys: Sequence of strings designating the PV label maps in
+            the sample that will be used to generate the new image.
+            Cannot be set at the same time as label_key.
+        image_key: String designating the key to which the new volume will be saved.
+            If this key corresponds to an already existing volume, zero elements from
+            the label maps will be filled with elements of the original volume.
+        gaussian_parameters: Dictionary containing the mean and standard deviation for
+            each label. For each value :math:`v`, if a tuple
+            :math:`(a, b)` is provided then
+            :math:`v \sim \mathcal{U}(a, b)`.
+            If no value is given for a label, value from default_gaussian_parameters
+            will be used.
+        default_gaussian_parameters: Dictionary containing the default
+            mean and standard deviation used for all labels that are not
+            defined in gaussian_parameters.
+            Default values are :math:`(0.1, 0.9)` for the mean and
+            :math:`(0.01, 0.1)` for the standard deviation.
+        binarize: Boolean to tell if PV label maps should be binarized.
+            Does not have any effects if not using PV label maps.
+            Binarization is done taking the highest value per voxel
+            in the different PV label maps.
```

Comment:
Why wouldn't you get 0s or 1s only?

Oh I think I understand your confusion, it should be written something like `"taking the class of the highest value per voxel"`. I guess with this correction it should be clearer but I'll explain anyways to make sure it is.

Let's start with the reason, in our case, we want to be able to binarize the PV maps to be able to quantify what using label maps vs using PV maps implies, what are the errors that we will make, etc... Binarizing the PV maps would then mean to get the "best" label map we can out of PV maps.

Now, how do we binarize the PV maps? For each voxel, we look at the different volumes from the PV maps and choose the argmax over the PV maps. This gives us a single value per voxel creating a label map. The only tricky thing here is to deal with "uncomplete" PV maps. Let's imagine that for example you only have maps of CSF, GM, WM. If you take the argmax over the voxels you will assign all your voxels to one of the three classes, that's why we had to add in the implementation to not use voxels where the max over the PV maps was equal to 0. (And by the way it is also these points that we can fill in with the original T1 or T2 image.)

Is this clearer?

Reply:
Yes, it's very clear, thanks! Especially useful is the last sentences explaining why the 0 voxels from the other image are useful. I guess my main point is that I think it's misleading to say binarize for this as, from my point of view, binarize means to make something into 0s and 1s and nothing else, which is not the case.

Reply:
Yes I agree. The word "binarize" makes sense if we see the PV label maps as different label maps because what we are doing is to transform each PV label map with values in range [0, 1] to a binary map. What should be the term when we consider the PV label maps as one unique map, "labelize"?

Reply:
_Discretize_?

> The word "binarize" makes sense if we see the PV label maps as different label maps because what we are doing is to transform each PV label map with values in range [0, 1] to a binary map.

I still don't understand this. This is what I assume happens if `binarize` is `True`:

```
[0.0 0.1 0.7 0.1]
[0.0 0.4 0.2 0.8]
[0.0 0.5 0.1 0.1]

[ x   2   0   1 ]
```

Where the final value at `x` would be taken from the original image. There are other possible values in the result apart from 0 and 1, so it's not binary.

Is that correct?

Reply:
> Discretize?

Fair enough 👍 

> Where the final value at x would be taken from the original image. There are other possible values in the result apart from 0 and 1, so it's not binary.

You are mixing two things there, the labeling and the image. What you are assuming is right, we'll get `[ x   2   0   1 ]` with `x` being `-1` (it's an implementation choice, it allows me to loop over positive values to find discretized the labels). Therefore you get a discretized label map with integer values ranging from `-1` to `nb_classes - 1`. 

Then, from your new label map you create your synthetic image by sampling a gaussian for each tissue or taking the values from the original image (corresponding to `-1` labels), creating non binary values but in the image, not in the label map.

Is that clearer?

Reply:
Yes, I understand the process. Just I never understood the reason to use the word binarize. But I guess that doesn't matter anymore if we settle to `discretize`. Maybe you can add the word `argmax` in the docstring of the argument, I think that'd help users understand its function.

Reply:
I guess I started using it as we were only dealing with a PV map of the grey matter and kept using it...

Reply:
hi there
I thing it is really a binarization the confusion comes from the fact that @GFabien switch between 4D hot encoding to 3D multiple labels
if I take the @fepegar  example the binarize result for PV example :

```
[0.0 0.1 0.7 0.1]
[0.0 0.4 0.2 0.8]
[0.0 0.5 0.1 0.1]

```
is 

```
[ 0 0 1 0 ]
[ 0 0 0 1 ]
[ 0 1 0 0 ]
```

Each line represent a specific tissue, and the operation is well a binarization. This is equivalent to a 3D multilabel where now the number codes for the specific tissue
[-1 2 0 1]

I think it is just an internal choice, although I do not fully understand why @GFabien choose the 3D multilabel, whereas keeping the 4D representation will give equivalent processing for PV and binarized PV
it woul avoid to have the if condition in 
https://github.com/fepegar/torchio/blob/3e928bb527168cf80ebc1abb934635e43d42297b/torchio/transforms/augmentation/intensity/random_labels_to_image.py#L163

does it makes sense ?

Reply:
> I think it is just an internal choice, although I do not fully understand why @GFabien choose the 3D multilabel, whereas keeping the 4D representation will give equivalent processing for PV and binarized PV

I agree, it would be much more elegant, the thing is that I tried to do it this way by making the label maps 4D using one hot encoding but it was hugely time consuming (I used the `torch.nn.functional.one_hot` function). Therefore I went back to this implementation which is less elegant but way less costly.

Reply:
Ok then it is just an internal tricks, fair enough
But just to be sure I understand,  if your PV map are binarized outside torchio, and then used without the args discretize, it will be keept in 4D, will it take more time ?

Reply:
I didn't run benchmarks but it should be the same thing as making operations on `n` 3D tensors compared to make one operation on a 4D tensor with `n` channels. Working with 3D tensors seems to be faster so if you discretize your labels outside of torchio you should probably give them as a 3D label map and not a one hot encoded one.